### PR TITLE
feat : created interfaces

### PR DESCRIPTION
**Pull request by sammychinedu2ky**
This tries to resolve this issue 
https://github.com/oras-project/oras-dotnet/issues/3
Yet to create the Copy type

### CODE REVIEW

File: OrasDotNet/Models/Descriptor.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+using System.Linq;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+
+
+namespace OrasDotnet.Models
+{
+    internal class Descriptor
```
Comment (Wwwsylvia) [inline OrasDotNet/Models/Descriptor.cs:12]: We need to make it public so that users can use it.

File: OrasDotNet/Models/DescriptorOCI.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+using System.Linq;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+
+
+namespace OrasDotnet.Models
+{
+    internal class DescriptorOCI
```
Comment (Wwwsylvia) [inline OrasDotNet/Models/DescriptorOCI.cs:12]: Wondering why is it called `DescriptorOCI`? How about `MinimumDescriptor`? This type should only be used internally.

File: OrasDotNet/Interfaces/IReadOnlyStorage.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿using OrasDotnet.Models;
+using System.Threading;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace OrasDotnet.Interfaces
+{
+    internal interface IReadOnlyStorage
```
Comment (Wwwsylvia) [inline OrasDotNet/Interfaces/IReadOnlyStorage.cs:8]:
```suggestion
    public interface IReadOnlyStorage
```

File: OrasDotNet/Interfaces/IStorage.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OrasDotnet.Interfaces
+{
+    internal interface IStorage : IReadOnlyStorage
```
Comment (Wwwsylvia) [inline OrasDotNet/Interfaces/IStorage.cs:12]: public

File: OrasDotNet/Interfaces/ITagResolver.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.Threading;
+using OrasDotnet.Models;
+
+namespace OrasDotnet.Interfaces
+{
+    internal interface ITagResolver
+    {
+        Task<Descriptor> ResolveAsync(string reference, CancellationToken cancellationToken = default);
```
Comment (Wwwsylvia) [inline OrasDotNet/Interfaces/ITagResolver.cs:13]: Should we have a `IResolver` for this?

File: OrasDotNet/Interfaces/ITagResolver.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        Task TagAsync(string reference, Descriptor descriptor, CancellationToken cancellationToken = default);
```
Comment (Wwwsylvia) [inline OrasDotNet/Interfaces/ITagResolver.cs:14]: nit: 
```suggestion
        Task TagAsync(Descriptor descriptor, string reference, CancellationToken cancellationToken = default);
```

To be consistent with `oras-go`.

File: OrasDotNet/Interfaces/ITarget.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OrasDotnet.Interfaces
+{
+    internal interface ITarget : IStorage, ITagResolver
```
Comment (Wwwsylvia) [inline OrasDotNet/Interfaces/ITarget.cs:7]: public.

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

**Comment by Wwwsylvia** [comment]
BTW should the title starts with "feat: "?

File: OrasDotNet/Interfaces/ITagResolver.cs (hunk 5 above)
Comment (sammychinedu2ky) [reply OrasDotNet/Interfaces/ITagResolver.cs:13]: From the mermaid diagram, I didn't really see a base a Resolver but only the IResolver

Comment (sammychinedu2ky) [reply OrasDotNet/Interfaces/ITagResolver.cs:13]: Ok no worries I've seen it in the GoLang code

File: OrasDotNet/Models/DescriptorOCI.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+using System.Linq;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+
+
+namespace OrasDotnet.Models
+{
+    internal class MinimumDescriptor
```
Comment (Wwwsylvia) [inline OrasDotNet/Models/DescriptorOCI.cs:12]: You may want to change the file name as well.

File: OrasDotNet/Models/Errors/AlreadyExistsException.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OrasDotNet.Models.Errors
+{
+    internal class AlreadyExistsException : Exception
```
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply OrasDotNet/Models/Errors/AlreadyExistsException.cs:7]: Sorry I don't get it😂 How do we customize the error message for this exception?

File: OrasDotNet/OrasDotNet.csproj
Comment (shizhMSFT) [inline OrasDotNet/OrasDotNet.csproj:1]: Should we just call this project `Oras`?

File: OrasDotNet/Models/DescriptorOCI.cs (hunk 8 above)
Comment (sammychinedu2ky) [reply OrasDotNet/Models/DescriptorOCI.cs:12]: ooh yh that's right

File: OrasDotNet/Interfaces/IReadOnlyStorage.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿using OrasDotnet.Models;
+using System.Threading;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace OrasDotnet.Interfaces
```
Comment (shizhMSFT) [inline OrasDotNet/Interfaces/IReadOnlyStorage.cs:6]: I'm thinking about the following namespaces
- `Oras.Content`
- `Oras.Registry`

and maybe
- `Oras.Exception`

File: OrasDotNet/Interfaces/IResolver.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using System.Threading;
+
+namespace OrasDotnet.Interfaces
+{
+    public interface IResolver
+    {
+        Task<Descriptor> ResolveAsync(string reference, CancellationToken cancellationToken = default);
+
```
Comment (shizhMSFT) [inline OrasDotNet/Interfaces/IResolver.cs:13]: nit: we should remove unnecessary empty lines.

File: OrasDotNet/Interfaces/IReadOnlyStorage.cs
Comment (shizhMSFT) [inline OrasDotNet/Interfaces/IReadOnlyStorage.cs:1]: We need to document all code

File: OrasDotNet/Models/Descriptor.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+        [JsonPropertyName("digest")]
+        public string Digest { get; set; }
+
+        [JsonPropertyName("size")]
+        public long Size { get; set; }
+
+        [JsonPropertyName("urls")]
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
+        public List<string> URLs { get; set; }
```
Comment (shizhMSFT) [inline OrasDotNet/Models/Descriptor.cs:26]: How about `IList`?
```suggestion
        public IList<string> URLs { get; set; }
```

File: OrasDotNet/Models/Descriptor.cs [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+        [JsonPropertyName("annotations")]
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
+        public Dictionary<string, string> Annotations { get; set; }
```
Comment (shizhMSFT) [inline OrasDotNet/Models/Descriptor.cs:30]: How about `IDictionary`
```suggestion
        public IDictionary<string, string> Annotations { get; set; }
```

File: OrasDotNet/OrasDotNet.csproj
Comment (sammychinedu2ky) [reply OrasDotNet/OrasDotNet.csproj:1]: Oras? hmm 🤔. That would be okay and short too.

**Comment by sammychinedu2ky** [comment]
Yh I think that would be a great idea

**Comment by sammychinedu2ky** [comment]
Yh I think that would be a great idea

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]
Should we change the directory name? Also please document all codes in comments.

File: Oras/Interfaces/IResolver.cs [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Oras.Interfaces
+{
+    /// <summary>
+    /// IResolver resolves reference tags.
+    /// </summary>
+    public interface IResolver
+    {
+        /// <summary>
+        /// ResolveAsync resolves the tag to a descriptor.
```
Comment (Wwwsylvia) [inline Oras/Interfaces/IResolver.cs:16]: nit:
```suggestion
        /// ResolveAsync resolves the reference to a descriptor.
```

**Review by Wwwsylvia — APPROVED** [review]
LGTM

**Comment by sammychinedu2ky** [comment]
> Please fix the DCO.

Thanks @Wwwsylvia .. its fixed now

File: Oras/Exceptions/InvalidReferenceException.cs [hunk 15]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Oras.Exceptions
+{
+    /// <summary>
+    /// InvalidDigestException is thrown when a digest is invalid.
```
Comment (Wwwsylvia) [inline Oras/Exceptions/InvalidReferenceException.cs:8]: This doc does not match the type.

**Pull request merged**